packages/mui-base/src/InputUnstyled/InputUnstyled.tsx

```diff
@@ -73,6 +73,10 @@ const InputUnstyled = React.forwardRef(function InputUnstyled(
     disabled,
     defaultValue,
     error,
+    inputRef:
+      typeof slotProps.input === 'object' && typeof slotProps.input.ref !== 'undefined'
+        ? slotProps.input.ref
+        : undefined,
```

Comment:
why do you need to pass ref from slotPros.input.ref? It will be forked in useSlotProps for Input. It is usefull to pass inputRef to useInput for developers who use this hook with custom components.

Reply:
> why do you need to pass ref from slotPros.input.ref?

If someone wants to provide their own ref to the InputUnstyled, they'd do it with the `slotPros.input.ref`.

> this is the best i could do to make typescript happy

Ah, I see, we've got a bit of a chicken and egg problem here - `slotProps` possibly need `ownerState`, which needs `useInput` result, which depend on `slotProps`. We have to do this differently, then. Do not pass `inputRef` to `useInput`, but add it to input slot's useSlotProps' `additionalProps` object. The useSlotProps hook should take care of merging it correctly with the result of useInput.

Reply:
So it turns out we need to pass `slotProps.input.ref` in additional props. `externalProps` in the same hook is taking care of passing `ref` to `useInput`.

Added test to make sure correct input element is attached to ref.

https://github.com/sai6855/material-ui/blob/79b43f90054c17382e596e4eee55034872d07611/packages/mui-base/src/InputUnstyled/InputUnstyled.test.tsx#L37

Reply:
Indeed! I forgot what functionality I added to useSlotProps ;)